switch dtensor and functional collective to use optree

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110670

optree recently landed and provide quite good perf, conditionally import
new optree if optree is installed

Some numbers testing mlp layer with TP + func collective:
before this PR: 10.390ms
after this PR: 9.189ms

so around e2e 10% CPU overhead reduction

wanchaol committed:
switch dtensor and functional collective to use optree

optree recently landed and provide quite good perf, conditionally import
new optree if optree is installed

[ghstack-poisoned]

fduwjj reviewed (commented):
wow do we know how much perf we can get from switching to optree?

wconstab commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -21,7 +21,12 @@
+try:
+    from torch.utils._cxx_pytree import tree_flatten, tree_unflatten
+except ImportError:
+    tree_flatten = torch.utils._pytree.tree_flatten  # type: ignore[assignment]
```
curious, why the assignment version here instead of doing import like above

wanchaol replied:
I followed a similar approach taken in https://github.com/pytorch/pytorch/pull/109684, probably another import also works (in both case we need to ignore type annotation, lmk your preference :)

wconstab commented on torch/distributed/_functional_collectives.py:
```diff
@@ -4,13 +4,18 @@
+try:
+    from torch.utils._cxx_pytree import tree_map_only
+except ImportError:
+    tree_map_only = torch.utils._pytree.tree_map_only  # type: ignore[assignment]
```
do we know under what builds/setups this import won't exist? Should we issue a warning if we're in a known slow mode?

wanchaol replied:
I believe some certain CI does not have access to `pip` channel and it only have conda channel, where optree not yet available in conda. @XuehaiPan might have more details

XuehaiPan replied:
I'm making `optree` optional in PR #109684. We don't need this try-except blocks after PR #109684.

- #109684

wanchaol commented:
> wow do we know how much perf we can get from switching to optree?

I tried it and there's not too much perf gain for large scale 2d, reason probably because we already hide almost all cpu overhead. This could potentially benefits smaller models we can try

fegin reviewed (approved):
Nice

XuehaiPan commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -334,7 +339,7 @@ class OpInfo:
     local_args: Sequence[object]
     local_kwargs: Dict[str, object]
-    args_tree_spec: Optional[TreeSpec] = None
+    args_tree_spec: Optional[PyTreeSpec] = None
```
We could leave this as `TreeSpec`. It is aliased to `PyTreeSpec` if `optree` is installed.

```suggestion
    args_tree_spec: Optional[TreeSpec] = None
```

XuehaiPan commented on torch/distributed/_tensor/op_schema.py:
```diff
@@ -4,7 +4,12 @@
 from torch.distributed._tensor.placement_types import DTensorSpec
-from torch.utils._pytree import tree_map_only, TreeSpec
+
+try:
+    from torch.utils._cxx_pytree import PyTreeSpec, tree_map_only
+except ImportError:
+    tree_map_only = torch.utils._pytree.tree_map_only  # type: ignore[assignment]
+    PyTreeSpec = torch.utils._pytree.TreeSpec  # type: ignore[assignment, misc]
```
```suggestion
try:
    from torch.utils._cxx_pytree import TreeSpec, tree_map_only
except ImportError:
    from torch.utils._pytree import TreeSpec, tree_map_only  # type: ignore[assignment]
```

wanchaol committed:
Update on "switch dtensor and functional collective to use optree"


optree recently landed and provide quite good perf, conditionally import
new optree if optree is installed

Some numbers testing mlp layer with TP + func collective:
before this PR: 10.390ms
after this PR: 9.189ms

so around e2e 10% CPU overhead reduction


[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
